salt/_pillar/metalk8s_nodes.py

```diff
@@ -50,13 +53,100 @@ def node_info(node, ca_minion):
     return result
 
 
+def get_cluster_version(api_client):
+    coreV1 = kubernetes.client.CoreV1Api(api_client=api_client)
+    try:
+        namespace = coreV1.read_namespace(name="kube-system")
+    except (ApiException, HTTPError) as exc:
+        return __utils__['pillar_utils.errors_to_dict'](
+            'Unable to read namespace information {}'.format(exc)
+        )
+    annotation_key = 'metalk8s.scality.com/cluster-version'
+
+    if annotation_key not in namespace.metadata.annotations:
+        return __utils__['pillar_utils.errors_to_dict'](
+            'Unable to retrieve cluster version'
+        )
+
+    return namespace.metadata.annotations[annotation_key]
+
+
+def iso_timestamp_converter(timestamp):
+    if not timestamp:
+        return None
+    return timestamp.isoformat()
+
+
+def get_storage_classes(api_client):
+    storageV1 = kubernetes.client.StorageV1Api(api_client=api_client)
+    storage_classes = {}
+    try:
+        storageclass_list = storageV1.list_storage_class()
+    except (ApiException, HTTPError) as exc:
+        log.exception('Exception when calling storageV1->list_storage_class')
+        return None
+    for storageclass in storageclass_list.items:
+        # Need to convert the datetime object in storageclass to ISO format in
+        # order to make them serializable.
+        storageclass.metadata.creation_timestamp = iso_timestamp_converter(
+            storageclass.metadata.creation_timestamp
+        )
+        storageclass.metadata.deletion_timestamp = iso_timestamp_converter(
+            storageclass.metadata.deletion_timestamp
+        )
+        storage_classes[storageclass.metadata.name] = storageclass.to_dict()
+
+    return storage_classes
+
+
+def list_volumes(api_client, minion_id):
+    results = {}
+    customObjectApi = kubernetes.client.CustomObjectsApi(api_client=api_client)
+    storage_classes = get_storage_classes(api_client)
+
+    if not storage_classes:
+        return {}
+
+    try:
+        volumes = customObjectApi.list_cluster_custom_object(
+            group="storage.metalk8s.scality.com",
+            version="v1alpha1",
+            plural="volumes"
+        )
+    except (ApiException, HTTPError) as exc:
+        return __utils__['pillar_utils.errors_to_dict'](
+            'Unable to get volumes {}'.format(exc)
+        )
+
+    local_volumes = (
+        volume for volume in volumes['items']
+        if volume['spec']['nodeName'] == minion_id
+    )
+    for volume in local_volumes:
+        name = volume['metadata']['name']
+        storageclass = storage_classes.get(
+            volume['spec']['storageClassName']
+        )
+        if storageclass:
```

Comment:
```python
        if storageclass:
            volume['spec']['storageClassName'] = storageclass
            results[name] = result
        else:
            log.error('Invalid storageclass for {} volume'.format(name))
```

I would get rid of the `result` var that just make things confusing w.r.t. `results` and simply skip the invalid volume instead of adding `None` (so that the SLS doesn't have to filter anything).

Reply:
@NicolasT as a baseline requirement, should we stick to listing all volumes in the pillar even if some volumes do not have storage classes?

Does it become problematic filtering out None in SLS files versus filtering out Objects for some reason? Because initially, we were returning strings for volumes with no storage class and I think you did mention that `None` should be used instead.

Reply:
Either not list them (but then log proper errors, as @slaperche-scality menioned), or `None`, but not strings. Checking for `None` in an SLS is much easier than checking or 'an object that looks like the kind of object I expect'.

However.

I guess @slaperche-scality is right and we should skip-but-log 'invalid' volumes and not represent them in the pillar.

The 'error' object should only be used when we can't properly render the pillar, i.e. there's a configuration issue or something along those lines.